File: webui/code/llm/chatlog.js

export function chatEntry(role, content) {
    return { role, content };
}

function noCollect() {
    return {
        name: null,
        collect: (line) => {},
        finalize: () => {}
    };
}

function literalCollector(name) {
    const lines = [];

    const collect = (line) => {
        lines.push(line);
    };

    const finalize = () => {
        let result = "";
        let inBlankRun = false;

        for (const line of lines) {
            if (line === "") {
                inBlankRun = true;
            } else {
                if (result.length > 0) {
                    result += inBlankRun ? '\n\n' : ' ';
                }
                result += line;
                inBlankRun = false;
            }
        }

        return result;
    };

    return {
        name,
        collect,
        finalize
    };
}

function chatCollector(name) {
    let chat = [];
    let collector = noCollect();

    const collect = (line) => {
        if (['.user:', '.system:', '.assistant:'].includes(line)) {
            let role = line.slice(1, -1);
            if (collector.name !== null) {
                chat.push(chatEntry(collector.name, collector.finalize()));
            }
            collector = literalCollector(role);
        }
        else {
            collector.collect(line);
        }
    };

    const finalize = () => {
        if (collector.name !== null) {
            chat.push(chatEntry(collector.name, collector.finalize()));
        }
        return chat;
    };

    return {
        name,
        collect,
        finalize
    };
}



export function readChatlog(text) {
    const lines = text.split('\n');

    if (
        lines.length < 3 || lines[0] !== '' ||
        !lines[1].startsWith('chatlog ') || lines[2] !== ''
    ) {
        throw new Error('invalid chatlog header');
    }

    // ignore version number for now

    let chatlog = {};

    let collector = noCollect();

    for (let lineIndex = 3; lineIndex < lines.length; lineIndex += 1) {
        const line = lines[lineIndex];

        if (line.startsWith(".chat ")) {
            let name = line.slice(6); // rest of the line
            if (collector.name !== null) {
                console.log(collector);
                chatlog[collector.name] = collector.finalize();
            }
            collector = chatCollector(name);
        }
        else if (line.startsWith(".literal ")) {
            let name = line.slice(9); // rest of the line
            if (collector.name !== null) {
                chatlog[collector.name] = collector.finalize();
            }
            collector = literalCollector(name);
        }
        else {
            collector.collect(line);
        }
    }

    if (collector.name !== null) {
        chatlog[collector.name] = collector.finalize();
    }

    return chatlog;
}

export async function loadChatlog(uri) {
    const response = await fetch(uri);

    if (!response.ok) {
        throw new Error(`could not load ${uri}`);
    }

    const source = await response.text();

    return readChatlog(source);
}

File: lib/llm/chatlog.py

import re
import json

from dataclasses import dataclass
from typing import List, Dict

from lib.core import AttrDict

def chatEntry(role: str, content: str) -> AttrDict:
    return AttrDict({"role": role, "content": content})

class Collector:
    def __init__(self, name=None):
        self.name = name

    def collect(self, line):
        pass

    def finalize(self):
        pass

class LiteralCollector(Collector):
    def __init__(self, name):
        super().__init__(name)
        self.lines = []

    def collect(self, line):
        self.lines.append(line)

    def finalize(self):
        result = ""
        in_blank_run = False
        for line in self.lines:
            if line == "":
                in_blank_run = True
            else:
                if result:
                    result += '\n\n' if in_blank_run else ' '
                result += line
                in_blank_run = False
        return result

class ChatCollector(Collector):
    def __init__(self, name):
        super().__init__(name)
        self.chat = []
        self.current_collector = Collector()

    def collect(self, line):
        if line in ['.user:', '.system:', '.assistant:']:
            role = line[1:-1]
            if self.current_collector.name is not None:
                self.chat.append(chatEntry(
                    self.current_collector.name,
                    self.current_collector.finalize()
                ))
            self.current_collector = LiteralCollector(role)
        else:
            self.current_collector.collect(line)

    def finalize(self):
        if self.current_collector.name is not None:
            self.chat.append(chatEntry(
                self.current_collector.name,
                self.current_collector.finalize()
            ))
        return self.chat

def apply_substitutions(source, substitutions: Dict):
    if isinstance(source, str):
        def replacer(match):
            key = match.group(1)
            if key not in substitutions:
                print(f"Substitution error, missing {key}.")
                return ""
            return json.dumps(substitutions[key], indent=4)
        return re.sub(r'\$\{\s*([^}]+?)\s*\}', replacer, source)
    elif isinstance(source, list):
        return [
            chatEntry(entry.role, apply_substitutions(entry.content, substitutions))
            for entry in source
        ]
    else:
        raise TypeError(f"source must be str or list of chatEntry, got {type(source)}")

def read_chatlog(text: str) -> Dict:
    lines = text.split('\n')
    if (len(lines) < 3 or lines[0] != '' or
        not lines[1].startswith('chatlog ') or lines[2] != ''):
        raise ValueError('invalid chatlog header')

    chatlog = {}
    collector = Collector()

    for line in lines[3:]:
        if line.startswith(".chat "):
            name = line[6:]
            if collector.name is not None:
                chatlog[collector.name] = collector.finalize()
            collector = ChatCollector(name)
        elif line.startswith(".literal "):
            name = line[9:]
            if collector.name is not None:
                chatlog[collector.name] = collector.finalize()
            collector = LiteralCollector(name)
        else:
            collector.collect(line)

    if collector.name is not None:
        chatlog[collector.name] = collector.finalize()

    return chatlog

def load_chatlog(filepath: str) -> AttrDict:
    with open(filepath, 'r') as f:
        return AttrDict(read_chatlog(f.read()))

